### PR TITLE
Bayem

### PR DESCRIPTION
**Pull request by TTitscher**
Solves #69

### CODE REVIEW

**Comment by TTitscher** [comment]
There are obvious flaws that I tried to address in e.g. #77 #76 , but those are IMO also present in the current `master`. So, after some review, I like to merge this ASAP and continue working using this new API.

**Comment by ajafarihub** [comment]
As all the tests have been passed, the merge is OK from my point of view.

File: bayem/json_io.py
Comment (joergfunger) [inline bayem/json_io.py:49]: Could you document what is stored in this __dict__ object, I found this (The namespace supporting arbitrary function attributes), but on a first glanced I'm confused why this not covered by the default.

File: bayem/vba.py
Comment (joergfunger) [inline bayem/vba.py:33]: Would it be possible to have that as a parameter (so e.g. in the constructor with standard e-7? In particular, it might be nice to be able to provide absolute values

File: bayem/vba.py
Comment (joergfunger) [inline bayem/vba.py:25]: how is that transformation defined?

File: bayem/vba.py
Comment (joergfunger) [inline bayem/vba.py:59]: isn't the not wrong? jac should be provided and callable=True?

File: bayem/vba.py
Comment (joergfunger) [inline bayem/vba.py:447]: Why do you store only the diagonal entries. I would say that this should rather be a user choice. The standard case could actually be to store the complete precision (that's a little more data, but you don't have to invert the matrix).

File: bayem/vba.py
Comment (joergfunger) [inline bayem/vba.py:472]: Regarding the discussion related to the ill-posed problem with C_beam and C_connection, I would argue that the std are not necessarily the interesting outputs, but maybe the eigenvalues/vectors of the posterior.

File: bayem/visualization.py
Comment (joergfunger) [inline bayem/visualization.py:102]: Maybe it would be nice to shortly document what is plotted (what is a diagonal plot?)

File: tests/test_vb.py
Comment (joergfunger) [inline tests/test_vb.py:2]: just out of curiosity, why did you switch from unittest to pytest?

File: tests/test_vb.py
Comment (joergfunger) [inline tests/test_vb.py:22]: could you test that automatically (if method is implemented, call it, otherwise don't?)

File: tests/test_vb_ARD.py
Comment (joergfunger) [inline tests/test_vb_ARD.py:2]: Is the ARD test case now working (I mean really giving reasonable results, not that the tests return True)?

**Review by joergfunger — CHANGES_REQUESTED** [review]
Nice piece of work. There are a few remarks from my side, but this interface is well structured and we could easily incorporate this in probeye. In particular, this solves the challenge that on the one hand we have a simple interface (given here), and if you need other solvers, then attach to probeye.

File: bayem/distributions.py
Comment (TTitscher) [reply bayem/distributions.py]: Added doc

File: bayem/distributions.py
Comment (TTitscher) [reply bayem/distributions.py:87]: Add pdf formula for Gamma and MVN

File: bayem/json_io.py
Comment (TTitscher) [reply bayem/json_io.py:49]: https://docs.python.org/3/library/stdtypes.html#object.__dict__: 
> A dictionary or other mapping object used to store an object’s (writable) attributes.

So it basically contains all the attributes of the object as a dict. But IMO this is well documented in python itself and the explanation does not belong in here. I also added a stackoverflow link above for further reading. But let me know, if you think otherwise. No hard feelings about that.

File: bayem/vba.py
Comment (TTitscher) [reply bayem/vba.py]: I added a `cdf_eps` to the `Options` class. 

Also note that this central differences Jacobian idea is IMO a reasonable default value for the _lazy_ user ;)
There are many ways to improve the Jacobian and the user can conveniently provide his/her own implementation. Would you agree?

File: bayem/vba.py
Comment (TTitscher) [reply bayem/vba.py]: Add docstring with explanation.

File: bayem/vba.py
Comment (TTitscher) [reply bayem/vba.py:79]: `jac_in_f` means that the call to `f` already contains the Jacobian, e.g.

~~~py
def my_model_error(theta):
   x = theta[0]
   return (time_steps * x**2, time_steps * 2*x)
~~~

In that case, the method is called with
~~~py
bayem.vba(my_model_error, x0, jac=True)
~~~

File: bayem/vba.py
Comment (TTitscher) [reply bayem/vba.py:475]: Add `store_full_precision=True` option.

File: bayem/vba.py
Comment (TTitscher) [reply bayem/vba.py]: Good point. However this method is inspired by the `summary` methods of other Bayesian inference frameworks, where I have not yet seen such your proposal. That could be something for the expert user to implement...

File: tests/test_vb.py
Comment (TTitscher) [reply tests/test_vb.py:2]: It is just more verbose:
* no subclassing of `unittest.TestCase` saves some LOC
* python-esque checks like `assert a == b` instead of `self.assertEqual(a, b)`

File: tests/test_vb.py
Comment (TTitscher) [reply tests/test_vb.py:22]: Your idea would also work. I don't know what is better and would leave it as is, to save some work/time :)

File: tests/test_vb_ARD.py
Comment (TTitscher) [reply tests/test_vb_ARD.py:2]: I'm not fully sure. The checks fail, if you remove the `index_ARD` option. #79 continues to investigate it.

File: bayem/distributions.py
Comment (ajafarihub) [inline bayem/distributions.py:32]: Minor:
maybe to use:
~~~py
self.parameter_names = [f"self.name-{i}" for i in range(len(self.mean))]
~~~

File: bayem/distributions.py
Comment (ajafarihub) [inline bayem/distributions.py:132]: I did not get what this cls is and aims for?

Comment (ajafarihub) [inline bayem/distributions.py:132]: Just FYI:
I also implemented this in a different way [here](https://git.bam.de/mechanics/ajafari/fenics-nonlocal-damage/-/blob/master/calibrates/fwme/helper_parameters.py#L303).

File: bayem/distributions.py
Comment (ajafarihub) [inline bayem/distributions.py:181]: Cool method !

File: bayem/vba.py
Comment (ajafarihub) [inline bayem/vba.py:87]: What transformation is meant here (and also for the next line of the code: self._TJ)? Is it by default no (identity) transformation?

File: bayem/vba.py
Comment (ajafarihub) [inline bayem/vba.py:365]: Minor:
I like to have it as a separate method. Would it probably make any better sense, if we have it as a class (static) method instead (just rough idea, as I myself implemented once before this way)?

File: bayem/distributions.py
Comment (TTitscher) [reply bayem/distributions.py:132]: I think you took the algorithm directly from the link in the docstring, right? I only did some adjustments and performance improvements.

File: bayem/vba.py
Comment (TTitscher) [reply bayem/vba.py:243]: I renamed the class to `DictModelError` and added a docstring with more information. See, if that is clearer now.

File: bayem/vba.py
Comment (TTitscher) [reply bayem/vba.py:420]: Really good point. I left that out, because that free method would have over 10 parameters and something inside me refuses to write such a method. I suggest to leave it for now and discuss a remedy in #80. OK?

File: bayem/distributions.py
Comment (TTitscher) [reply bayem/distributions.py:32]: Well, according to the paper notation, those would better be called `θ_{i}` or `$\theta_{i}$`. Could you live with that, what would you prefer?

Comment (TTitscher) [reply bayem/distributions.py]: The `\theta` latex thing looks _much_ better in the provided `visualization` methods. So I used that for now. Are you fine with that?

File: bayem/json_io.py
Comment (joergfunger) [reply bayem/json_io.py:50]: yes, for me it is sufficient have a link for someone not as profound in python as me.

File: bayem/vba.py
Comment (joergfunger) [reply bayem/vba.py]: Maybe this could be stressed here, the basic user might not realize how to change that himself.

**Review by joergfunger — APPROVED** [review]
Apart from some really minor open comments I suggest to merge. Thanks

File: bayem/distributions.py
Comment (ajafarihub) [reply bayem/distributions.py]: Yes sure. As I wrote in my comment. It was just a minor thing (suggestion).

File: bayem/vba.py
Comment (ajafarihub) [reply bayem/vba.py:243]: I see now. I had thought , it would be related to transformation of the qunatities (numbers themselves) , but it is only about the structure for representing them. Alright, is now clear.

**Comment by ajafarihub** [comment]
Thanks @TTitscher  for all changes and improvements. I would also like to have it merged to the master as soon as possible.

**Pull request merged**
